In [1]:
import pandas as pd
import numpy as np
#Método de clasificación KNN (vecino más cercano)
from sklearn.neighbors import KNeighborsClassifier
#Método de clasifiación DecisionTree (árbol de decision)
from sklearn.tree import DecisionTreeClassifier
#Uso de gráficas y representación de datos
import matplotlib.pyplot as plt
from sklearn import datasets
from scipy.io import arff


In [2]:
#Lectura de Dataset haciendo uso de URL de github y el nombre de las columnas
def readDataset(file,columns):
  dataset=pd.read_csv(file, header=None, names=columns)
  return dataset

###**1. Seleccione al menos tres conjuntos de datos de 2 classes. Es preferible conjuntos pequeños con variables nominales.**###

In [3]:
#Diabetes Data Set
columnas=['Times Pregnant','Glucose tolerance','Blood pressure','Triceps skin fold','Insulin','Body mass','Diabetes pedigree', 'Age','Class']
diabetes_features=['Times Pregnant','Glucose tolerance','Blood pressure','Triceps skin fold','Insulin','Body mass','Diabetes pedigree', 'Age']
diabetes=readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/diabetes.data',columnas)
diabetes_X=diabetes[diabetes_features]
diabetes_y=diabetes['Class']
diabetes.head()

Times Pregnant  Glucose tolerance  Blood pressure  Triceps skin fold  \
0               6                148              72                 35   
1               1                 85              66                 29   
2               8                183              64                  0   
3               1                 89              66                 23   
4               0                137              40                 35   

   Insulin  Body mass  Diabetes pedigree  Age            Class  
0        0       33.6              0.627   50  tested_positive  
1        0       26.6              0.351   31  tested_negative  
2        0       23.3              0.672   32  tested_positive  
3       94       28.1              0.167   21  tested_negative  
4      168       43.1              2.288   33  tested_positive

In [4]:
#Vote Data set
columnas = ['handicapped-infants','water cost sharing', 'adoption the budget resolution', 'physician fee freezen', 'god aid','religous groups','anti satellite test','aid to nicaraguan contras','mx-misil','inmigration','recortes','gasto de educacion','demandas','crimen','exportacion libre impuestos','adminitracion sur africa', 'class']
vote_features = ['handicapped-infants','water cost sharing', 'adoption the budget resolution', 'physician fee freezen', 'god aid','religous groups','anti satellite test','aid to nicaraguan contras','mx-misil','inmigration','recortes','gasto de educacion','demandas','crimen','exportacion libre impuestos','adminitracion sur africa']
vote = readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/vote.data',columnas)
vote = vote.replace({'y':0, 'n':1, '?':2})
vote_X = vote[vote_features]
vote_y = vote['class']
vote.head()

handicapped-infants  water cost sharing  adoption the budget resolution  \
0                    1                   0                               1   
1                    1                   0                               1   
2                    2                   0                               0   
3                    1                   0                               0   
4                    0                   0                               0   

   physician fee freezen  god aid  religous groups  anti satellite test  \
0                      0        0                0                    1   
1                      0        0                0                    1   
2                      2        0                0                    1   
3                      1        2                0                    1   
4                      1        0                0                    1   

   aid to nicaraguan contras  mx-misil  inmigration  recortes  \
0                          1         1            0         2   
1                          1         1            1         1   
2                          1         1            1         0   
3                          1         1            1         0   
4                          1         1            1         0   

   gasto de educacion  demandas  crimen  exportacion libre impuestos  \
0                   0         0       0                            1   
1                   0         0       0                            1   
2                   1         0       0                            1   
3                   1         0       1                            1   
4                   2         0       0                            0   

   adminitracion sur africa       class  
0                         0  republican  
1                         2  republican  
2                         1    democrat  
3                         0    democrat  
4                         0    democrat

In [37]:
#Titanic Data set
titanic = pd.read_csv('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/titanic.csv')
titanic=titanic.dropna()
titanic = titanic.replace({'male':0, 'female':1})
titanic['Survived'] = titanic['Survived'].replace({0:'NoSurvived',1:'Survived'})
titanic_X=titanic.drop(['Survived'],axis=1)
titanic_y=titanic['Survived']
titanic.head()

PassengerId    Survived  Pclass  \
1             2    Survived       1   
3             4    Survived       1   
6             7  NoSurvived       1   
10           11    Survived       3   
11           12    Survived       1   

                                                 Name  Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0      1   
6                             McCarthy, Mr. Timothy J    0  54.0      0   
10                    Sandstrom, Miss. Marguerite Rut    1   4.0      1   
11                           Bonnell, Miss. Elizabeth    1  58.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
3       0    113803  53.1000  C123        S  
6       0     17463  51.8625   E46        S  
10      1   PP 9549  16.7000    G6        S  
11      0    113783  26.5500  C103        S

###**2. Para cada uno de estos conjuntos aplique el método RIPPER del módulo “wittgenstein” (ejemplo en ripper.py). Evalúe los modelos obtenidos y compárelos con un árbol de decisión. Trate de interpretar los modelos obtenidos.**###

In [13]:
!pip install wittgenstein

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.8/124.8 KB 3.7 MB/s eta 0:00:00


In [14]:
import pandas as pd
from sklearn import datasets
import wittgenstein as lw
from sklearn.model_selection import train_test_split

In [16]:
train_diabetes, test_diabetes = train_test_split(diabetes, test_size=.33)

In [24]:
# Create and train model
ripper_clf = lw.RIPPER()
ripper_clf.fit(diabetes, class_feat="Class", pos_class='tested_positive')

# Print model
print(ripper_clf.out_model())

# Score
X_test_diabetes = test_diabetes.drop('Class', axis=1)
y_test_diabetes = test_diabetes['Class']
print(ripper_clf.score(X_test_diabetes, y_test_diabetes))

[[Glucosetolerance=>167.0] V
[Age=42.6-51.0 ^ Insulin=>210.0] V
[Glucosetolerance=147.0-167.0 ^ Diabetespedigree=0.37-0.45] V
[Age=29.0-33.0 ^ Insulin=150.0-210.0] V
[Tricepsskinfold=<8.2 ^ Glucosetolerance=134.0-147.0 ^ Bloodpressure=<54.0] V
[Bodymass=>41.5 ^ Bloodpressure=>88.0] V
[Age=38.0-42.6 ^ Glucosetolerance=109.0-117.0]]
None
0.7598425196850394


In [21]:
train_vote, test_vote= train_test_split(vote, test_size=.33)

In [29]:
# Create and train model
ripper_clf = lw.RIPPER()
ripper_clf.fit(vote, class_feat="class", pos_class='democrat')

# Print model
print(ripper_clf.out_model())

# Score
X_test_vote = test_vote.drop('class',axis=1)
y_test_vote = test_vote['class']
print(ripper_clf.score(X_test_vote, y_test_vote))

[[physicianfeefreezen=1 ^ adoptionthebudgetresolution=0] V
[physicianfeefreezen=1] V
[physicianfeefreezen=2 ^ watercostsharing=0] V
[recortes=0 ^ mx-misil=0] V
[recortes=0 ^ adoptionthebudgetresolution=0 ^ antisatellitetest=1] V
[adminitracionsurafrica=2 ^ adoptionthebudgetresolution=0 ^ antisatellitetest=1]]
None
0.9444444444444444


In [38]:
train_titanic, test_titanic = train_test_split(titanic, test_size=.33)

In [40]:
# Create and train model
ripper_clf = lw.RIPPER()
ripper_clf.fit(titanic, class_feat="Survived", pos_class='Survived')

# Print model
print(ripper_clf.out_model())

# Score
X_test_titanic = test_titanic.drop('Survived',axis=1)
y_test_titanic = test_titanic['Survived']
print(ripper_clf.score(X_test_titanic, y_test_titanic))

[[Sex=1] V
[Age=<17.2]]
None
0.7704918032786885


Este código está creando un modelo de clasificación utilizando el algoritmo RIPPER (Repeated Incremental Pruning to Produce Error Reduction) y entrenándolo con los datos del conjunto de datos titanic. La columna "Survived" se establece como la característica de clase y la clase positiva es "survived". Luego se imprime el modelo entrenado y se calcula la precisión del modelo utilizando los datos de prueba.

Este es el resultado esperado. La salida del método "out_model()" es una representación en forma de reglas del modelo entrenado. El modelo está haciendo una predicción basada en dos características: "Sex" y "Age".
En cuanto al None, es probable que sea debido a que el método out_model() no devuelve ningún valor, sólo imprime la salida.
Por último, el número 0.7704918032786885 es el puntaje de precisión del modelo en el conjunto de pruebas. Este puntaje indica que el modelo es capaz de predecir correctamente el 77% de las veces.





###**3. Seleccione al menos 10 conjuntos de datos. Genere un clasificador SVM con un kernel lineal y un valor fijo de C=1. Compare el rendimiento con un árbol de decisión**###

In [45]:
#Método de clasifiación DecisionTree (árbol de decision)
from sklearn.tree import DecisionTreeClassifier
#Método de clasificación SVM (Máquina de vector soporte)
from sklearn.svm import SVC
#Métrica para medir los métodos de clasifiacion
from sklearn import metrics

In [46]:
#Modelos a utilizar en esta práctica
tree=DecisionTreeClassifier()
svm=SVC(kernel='linear', C=1)

In [47]:
#Compute accuracy of model
def accuracyDataset(features, label, model): 
  accuracy=0
  #Split into random train and test subsets
  X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.2, random_state=1)
  #Entrenamiento de datos de entrenamiento
  model.fit(X_train,y_train)
  #Predice salidas de los datos test
  y_pred = model.predict(X_test)
  #Calcula precisión del modelo
  accuracy = metrics.accuracy_score(y_test, y_pred)
  return accuracy

In [49]:
#Iris Data set
columnas=['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)', 'class']
iris_features=['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
iris= readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/iris.data', columnas)
iris_X= iris[iris_features]
iris_y= iris['class']
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

         class  
0  Iris-setosa  
1  Iris-setosa  
2  Iris-setosa  
3  Iris-setosa  
4  Iris-setosa

In [52]:
iris_tree=accuracyDataset(iris_X,iris_y, tree)
iris_tree

0.9666666666666667

In [53]:
iris_svm=accuracyDataset(iris_X,iris_y, svm)
iris_svm

1.0

In [55]:
#Car Evaluation Data set
columnas=['buying','maint','doors', 'persons', 'lug_boot','safety','class']
car_features=['buying','maint','doors', 'persons', 'lug_boot','safety']
car= readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/car.data',columnas)
car['buying']=car['buying'].map({'vhigh':4, 'high':3, 'med':2, 'low':1})
car['maint']=car['maint'].map({'vhigh':4, 'high':3, 'med':2, 'low':1})
car['lug_boot']=car['lug_boot'].map({'small':1, 'med':2, 'big':3})
car['safety']=car['safety'].map({'low':1, 'med':2, 'high':3})
car['doors']=car['doors'].map({'1':1, '2':2,'3':3,'4':4,'5more':7}) #Esto hará que 7 signifique más de 5 puertas o personas
car['persons']=car['persons'].map({'1':1, '2':2,'3':3,'4':4,'more':7})
car_X=car[car_features]
car_y=car['class']
car.head()

buying  maint  doors  persons  lug_boot  safety  class
0       4      4      2        2         1       1  unacc
1       4      4      2        2         1       2  unacc
2       4      4      2        2         1       3  unacc
3       4      4      2        2         2       1  unacc
4       4      4      2        2         2       2  unacc

In [56]:
car_tree=accuracyDataset(car_X,car_y, tree)
car_tree

0.976878612716763

In [57]:
car_svm = accuracyDataset(car_X,car_y, svm)
car_svm

0.8323699421965318

In [59]:
#Wine Data Set
columnas= ['class','alcohol', 'malic acid', 'ash', 'alcalinityOfAsh', 'Magnesium',' Total Phenols', 'Flavanoids', 'Nonflavanoid phenols', 'Proanthocyanins', 'Color Intensity', 'Hue', 'OD280',' Proline'] 
wine_features=['alcohol', 'malic acid', 'ash', 'alcalinityOfAsh', 'Magnesium',' Total Phenols', 'Flavanoids', 'Nonflavanoid phenols', 'Proanthocyanins', 'Color Intensity', 'Hue', 'OD280',' Proline'] 
wine=readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/wine.data',columnas)
wine_X=wine[wine_features]
wine_y=wine['class']
wine.head()

class  alcohol  malic acid   ash  alcalinityOfAsh  Magnesium  \
0      1    14.23        1.71  2.43             15.6        127   
1      1    13.20        1.78  2.14             11.2        100   
2      1    13.16        2.36  2.67             18.6        101   
3      1    14.37        1.95  2.50             16.8        113   
4      1    13.24        2.59  2.87             21.0        118   

    Total Phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0            2.80        3.06                  0.28             2.29   
1            2.65        2.76                  0.26             1.28   
2            2.80        3.24                  0.30             2.81   
3            3.85        3.49                  0.24             2.18   
4            2.80        2.69                  0.39             1.82   

   Color Intensity   Hue  OD280   Proline  
0             5.64  1.04   3.92      1065  
1             4.38  1.05   3.40      1050  
2             5.68  1.03   3.17      1185  
3             7.80  0.86   3.45      1480  
4             4.32  1.04   2.93       735

In [60]:
wine_tree=accuracyDataset(wine_X,wine_y,tree)
wine_tree

0.8888888888888888

In [61]:
wine_svm=accuracyDataset(wine_X,wine_y, svm)
wine_svm

0.9444444444444444

In [62]:
#Diabetes Data Set
columnas=['Times Pregnant','Glucose tolerance','Blood pressure','Triceps skin fold','Insulin','Body mass','Diabetes pedigree', 'Age','Class']
diabetes_features=['Times Pregnant','Glucose tolerance','Blood pressure','Triceps skin fold','Insulin','Body mass','Diabetes pedigree', 'Age']
diabetes=readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/diabetes.data',columnas)
diabetes_X=diabetes[diabetes_features]
diabetes_y=diabetes['Class']
diabetes.head()


Times Pregnant  Glucose tolerance  Blood pressure  Triceps skin fold  \
0               6                148              72                 35   
1               1                 85              66                 29   
2               8                183              64                  0   
3               1                 89              66                 23   
4               0                137              40                 35   

   Insulin  Body mass  Diabetes pedigree  Age            Class  
0        0       33.6              0.627   50  tested_positive  
1        0       26.6              0.351   31  tested_negative  
2        0       23.3              0.672   32  tested_positive  
3       94       28.1              0.167   21  tested_negative  
4      168       43.1              2.288   33  tested_positive

In [63]:
diabetes_tree=accuracyDataset(diabetes_X, diabetes_y, tree)
diabetes_tree

0.6818181818181818

In [64]:
diabetes_svm=accuracyDataset(diabetes_X,diabetes_y, svm)
diabetes_svm

0.7792207792207793

In [65]:
#Glass Data set
columnas=['RI','Na', 'Mg', 'Al', 'Si','K', 'Ca', 'Ba', 'Fe', 'Type']
glass_features=['RI','Na', 'Mg', 'Al', 'Si','K', 'Ca', 'Ba', 'Fe']
glass=readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/glass.data',columnas)
glass_X=glass[glass_features]
glass_y=glass['Type']
glass.head()

RI     Na    Mg    Al     Si     K     Ca   Ba    Fe  \
0  1.51793  12.79  3.50  1.12  73.03  0.64   8.77  0.0  0.00   
1  1.51643  12.16  3.52  1.35  72.89  0.57   8.53  0.0  0.00   
2  1.51793  13.21  3.48  1.41  72.64  0.59   8.43  0.0  0.00   
3  1.51299  14.40  1.74  1.54  74.55  0.00   7.59  0.0  0.00   
4  1.53393  12.30  0.00  1.00  70.16  0.12  16.19  0.0  0.24   

                     Type  
0      'build wind float'  
1      'vehic wind float'  
2      'build wind float'  
3               tableware  
4  'build wind non-float'

In [66]:
glass_tree=accuracyDataset(glass_X,glass_y,tree)
glass_tree

0.627906976744186

In [68]:
glass_svm=accuracyDataset(glass_X,glass_y,svm)
glass_svm

0.5813953488372093

In [71]:
#Breast Cancer Wisconsin (Diagnostic) Data set
columnas=['id', 'Clump Thickness', 'Cell Size', 'Cell Shape', 'Marginal Adhesion', 'Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin', 'Normal Nucleoli', 'Mitoses', 'Class']
cancer_features=['Clump Thickness', 'Cell Size', 'Cell Shape', 'Marginal Adhesion', 'Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin', 'Normal Nucleoli', 'Mitoses']
cancer=readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/breast-cancer-wisconsin.data', columnas)
cancer_X=cancer[cancer_features]
cancer_y= cancer['Class']
cancer.head()

id  Clump Thickness  Cell Size  Cell Shape  Marginal Adhesion  \
0  1000025                5          1           1                  1   
1  1002945                5          4           4                  5   
2  1015425                3          1           1                  1   
3  1016277                6          8           8                  1   
4  1017023                4          1           1                  3   

   Epithelial Cell Size  Bare Nuclei  Bland Chromatin  Normal Nucleoli  \
0                     2            1                3                1   
1                     7           10                3                2   
2                     2            2                3                1   
3                     3            4                3                7   
4                     2            1                3                1   

   Mitoses  Class  
0        1      2  
1        1      2  
2        1      2  
3        1      2  
4        1      2

In [72]:
cancer_tree=accuracyDataset(cancer_X,cancer_y,tree)
cancer_tree

0.9416058394160584

In [73]:
cancer_svm=accuracyDataset(cancer_X,cancer_y,svm)
cancer_svm

0.9854014598540146

In [74]:
#Titanic Data set
titanic = pd.read_csv('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/titanic.csv')
titanic=titanic.drop(['PassengerId','Name', 'Ticket','Cabin','Embarked'],axis=1)
titanic=titanic.dropna()
titanic = titanic.replace({'male':0, 'female':1})
titanic_X=titanic.drop(['Survived'],axis=1)
titanic_y=titanic['Survived']
titanic.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare
0         0       3    0  22.0      1      0   7.2500
1         1       1    1  38.0      1      0  71.2833
2         1       3    1  26.0      0      0   7.9250
3         1       1    1  35.0      1      0  53.1000
4         0       3    0  35.0      0      0   8.0500

In [75]:
titanic_tree=accuracyDataset(titanic_X,titanic_y, tree)
titanic_tree

0.8251748251748252

In [76]:
titanic_svm=accuracyDataset(titanic_X,titanic_y, svm)
titanic_svm

0.7692307692307693

In [84]:
#Flag DataSet
columnas= ['name','landmass', 'zone', 'area','population','language','religion','bars','stripes','colours','red','green','blue','gold','white','black','orange','mainhue','circles','crosses','saltires','quarters','sunstars','crescent','triangle','icon','animate','text','topleft','botright']
flag_features=['landmass', 'zone', 'area','population','language','religion','bars','stripes','colours','red','green','blue','gold','white','black','orange','circles','crosses','saltires','quarters','sunstars','crescent','triangle','icon','animate','text']
flag=readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/flag.data',columnas)
flag_X = flag[flag_features]
flag_y = flag['name']
flag.head()

name  landmass  zone  area  population  language  religion  bars  \
0     Afghanistan         5     1   648          16        10         2     0   
1         Albania         3     1    29           3         6         6     0   
2         Algeria         4     1  2388          20         8         2     2   
3  American-Samoa         6     3     0           0         1         1     0   
4         Andorra         3     1     0           0         6         0     3   

   stripes  colours  ...  saltires  quarters  sunstars  crescent  triangle  \
0        3        5  ...         0         0         1         0         0   
1        0        3  ...         0         0         1         0         0   
2        0        3  ...         0         0         1         1         0   
3        0        5  ...         0         0         0         0         1   
4        0        3  ...         0         0         0         0         0   

   icon  animate text  topleft  botright  
0     1        0    0    black     green  
1     0        1    0      red       red  
2     0        0    0    green     white  
3     1        1    0     blue       red  
4     0        0    0     blue       red  

[5 rows x 30 columns]

In [85]:
flag_tree=accuracyDataset(flag_X,flag_y, tree)
flag_tree

0.0

In [87]:
flag_svm=accuracyDataset(flag_X,flag_y,svm)
flag_svm

0.0

In [86]:
#Vote Data set
columnas = ['handicapped-infants','water cost sharing', 'adoption the budget resolution', 'physician fee freezen', 'god aid','religous groups','anti satellite test','aid to nicaraguan contras','mx-misil','inmigration','recortes','gasto de educacion','demandas','crimen','exportacion libre impuestos','adminitracion sur africa', 'class']
vote_features = ['handicapped-infants','water cost sharing', 'adoption the budget resolution', 'physician fee freezen', 'god aid','religous groups','anti satellite test','aid to nicaraguan contras','mx-misil','inmigration','recortes','gasto de educacion','demandas','crimen','exportacion libre impuestos','adminitracion sur africa']
vote = readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/vote.data',columnas)
vote = vote.replace({'y':0, 'n':1, '?':2})
vote_X = vote[vote_features]
vote_y = vote['class']
vote.head()

handicapped-infants  water cost sharing  adoption the budget resolution  \
0                    1                   0                               1   
1                    1                   0                               1   
2                    2                   0                               0   
3                    1                   0                               0   
4                    0                   0                               0   

   physician fee freezen  god aid  religous groups  anti satellite test  \
0                      0        0                0                    1   
1                      0        0                0                    1   
2                      2        0                0                    1   
3                      1        2                0                    1   
4                      1        0                0                    1   

   aid to nicaraguan contras  mx-misil  inmigration  recortes  \
0                          1         1            0         2   
1                          1         1            1         1   
2                          1         1            1         0   
3                          1         1            1         0   
4                          1         1            1         0   

   gasto de educacion  demandas  crimen  exportacion libre impuestos  \
0                   0         0       0                            1   
1                   0         0       0                            1   
2                   1         0       0                            1   
3                   1         0       1                            1   
4                   2         0       0                            0   

   adminitracion sur africa       class  
0                         0  republican  
1                         2  republican  
2                         1    democrat  
3                         0    democrat  
4                         0    democrat

In [88]:
vote_tree=accuracyDataset(vote_X,vote_y,tree)
vote_tree

0.9655172413793104

In [89]:
vote_svm=accuracyDataset(vote_X,vote_y,svm)
vote_svm

0.9655172413793104

In [91]:
#Segment_challeng
columnas = ['region-centroid-col','region-centroid-row','region-pixel-count','short-line-density-5','short-line-density-2','vedge-mean','vegde-sd','hedge-mean','hedge-sd','intensity-mean','rawred-mean','rawblue-mean','rawgreen-mean','exred-mean','exblue-mean','exgreen-mean','value-mean','saturation-mean','hue-mean','class']
segment_features = ['region-centroid-col','region-centroid-row','region-pixel-count','short-line-density-5','short-line-density-2','vedge-mean','vegde-sd','hedge-mean','hedge-sd','intensity-mean','rawred-mean','rawblue-mean','rawgreen-mean','exred-mean','exblue-mean','exgreen-mean','value-mean','saturation-mean','hue-mean']
segment = readDataset('https://raw.githubusercontent.com/victoriapl01/Mineria_De_Datos/main/Datasets/segment_challeng.data',columnas)
segment_X = segment[segment_features]
segment_y = segment['class']
segment.head()

region-centroid-col  region-centroid-row  region-pixel-count  \
0                  144                   35                   9   
1                  118                  180                   9   
2                    6                  174                   9   
3                  152                  220                   9   
4                  189                  142                   9   

   short-line-density-5  short-line-density-2  vedge-mean  vegde-sd  \
0                   0.0                   0.0    2.333330  2.033060   
1                   0.0                   0.0    1.944440  1.481990   
2                   0.0                   0.0    1.888890  1.007410   
3                   0.0                   0.0    0.944445  0.685185   
4                   0.0                   0.0    0.000000  0.000000   

   hedge-mean  hedge-sd  intensity-mean  rawred-mean  rawblue-mean  \
0     2.05556   1.73098         37.5926      32.3333       47.4444   
1     3.11111   1.08866         48.5556      44.1111       59.0000   
2     2.88889   4.02963         19.0741      15.1111       17.7778   
3     1.44444   2.16296         14.6296      11.5556       13.1111   
4     0.00000   0.00000          0.0000       0.0000        0.0000   

   rawgreen-mean  exred-mean  exblue-mean  exgreen-mean  value-mean  \
0        33.0000   -15.77780     29.55560      -13.7778     47.4444   
1        42.5556   -13.33330     31.33330      -18.0000     59.0000   
2        24.3333   -11.88890     -3.88889       15.7778     24.3333   
3        19.2222    -9.22222     -4.55556       13.7778     19.2222   
4         0.0000     0.00000      0.00000        0.0000      0.0000   

   saturation-mean  hue-mean   class  
0         0.319714  -2.13876  cement  
1         0.278822  -1.99604    path  
2         0.381867   2.39502   grass  
3         0.416705   2.30688   grass  
4         0.000000   0.00000  window

In [93]:
segment_tree=accuracyDataset(segment_X,segment_y,tree)
segment_tree

0.9444444444444444

In [94]:
segment_svm=accuracyDataset(segment_X,segment_y,svm)
segment_svm

0.9629629629629629

###**4. Utilice el método GridSearchCV para obtener los mejores híper-parámetros usando los siguientes valores:**###

###**Set the parameters by cross-validation parameters = [    {"kernel": ["rbf"], "gamma": [0.01, 0.1, 1.0], "C": [1, 10, 100, 1000]},    {"kernel": ["linear"], "C": [1, 10, 100, 1000]}, ] NOTAS: Compare el rendimiento de los resultados usando validación cruzada de híper-parámetros y los métodos usando valores fijos**###

In [98]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn import svm
parameters = [ 
    {"kernel": ["rbf"], "gamma": [0.01, 0.1, 1.0], "C": [1, 10, 100, 1000]}, 
    {"kernel": ["linear"], "C": [1, 10, 100, 1000]}, ]
svm_grid= GridSearchCV(svm.SVC(), parameters, cv=5,scoring='accuracy')

In [101]:
iris_grid=svm_grid.fit(iris_X, iris_y)
print ('Mejor hiperparametros',iris_grid.best_params_)

Mejor hiperparametros {'C': 1, 'gamma': 0.1, 'kernel': 'rbf'}


In [102]:
car_grid=svm_grid.fit(car_X,car_y)
print ('Mejor hiperparametros',car_grid.best_params_)

Mejor hiperparametros {'C': 10, 'gamma': 0.1, 'kernel': 'rbf'}


In [103]:
wine_grid=svm_grid.fit(wine_X,wine_y)
print ('Mejor hiperparametros',wine_grid.best_params_)

In [104]:
diabetes_grid=svm_grid.fit(diabetes_X,diabetes_y)
print ('Mejor hiperparametros',diabetes_grid.best_params_)

Mejor hiperparametros {'C': 100, 'kernel': 'linear'}


In [105]:
glass_grid=svm_grid.fit(glass_X,glass_y)
print ('Mejor hiperparametros',glass_grid.best_params_)

Mejor hiperparametros {'C': 10, 'gamma': 1.0, 'kernel': 'rbf'}


In [106]:
cancer_grid=svm_grid.fit(glass_X,glass_y)
print ('Mejor hiperparametros',cancer_grid.best_params_)

Mejor hiperparametros {'C': 10, 'gamma': 1.0, 'kernel': 'rbf'}


In [107]:
titanic_grid=svm_grid.fit(titanic_X, titanic_y)
print ('Mejor hiperparametros',titanic_grid.best_params_)

Mejor hiperparametros {'C': 1000, 'kernel': 'linear'}


In [109]:
vote_grid=svm_grid.fit(vote_X,vote_y)
print ('Mejor hiperparametros',vote_grid.best_params_)

Mejor hiperparametros {'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}


In [110]:
segment_grid=svm_grid.fit(segment_X,segment_y)
print ('Mejor hiperparametros',segment_grid.best_params_)

Mejor hiperparametros {'C': 100, 'kernel': 'linear'}


La validación cruzada de hiperparámetros se realiza cuando se quiere seleccionar los mejores hiperparámetros para un modelo dado. Es una técnica utilizada para evaluar el rendimiento del modelo con diferentes combinaciones de hiperparámetros, con el objetivo de encontrar la combinación óptima que maximice el rendimiento del modelo.

Se suele realizar después de elegir el modelo, y antes de entrenarlo con el conjunto de entrenamiento completo y evaluarlo con el conjunto de pruebas. Es una forma de asegurar que el modelo seleccionado es el más adecuado para el conjunto de datos y que se establecen los mejores hiperparámetros posibles.

Además, es especialmente útil cuando el modelo tiene muchos hiperparámetros y es difícil saber cuáles son los mejores valores para estos parámetros. La validación cruzada de hiperparámetros permite automatizar este proceso de selección de hiperparámetros y ahorra tiempo y esfuerzo.

uí hay un ejemplo de código que utiliza la validación cruzada de hiperparámetros en Python con scikit-learn para seleccionar los mejores hiperparámetros para un modelo de clasificación de árboles de decisión:

GRIDSEARCHCV
Hacer gráfico de los resultados dados

COMPARACIÓN DEL PUNTO 3 Y EL PUNTO 4